File: Modelos Python/modelo/modelo.py
import random
import math

class Politico:
    def __init__(self, id, dineroRobado, edad, fechaNacimiento):
        self.id = id
        self.dineroRobado = dineroRobado
        self.edad = edad
        self.fechaNacimiento = fechaNacimiento

class Modelo:
    def __init__(self):
        self.arrayPoliticos = []
        self.matrizPoliticos = []
        self.k = 0
        self.m = 0

    def generar_array(self, n):
        self.arrayPoliticos.clear()
        for i in range(n):
            edad = random.randint(30, 79)
            p = Politico(i + 1, random.randint(100, 1000000), edad, str(2024 - edad))
            self.arrayPoliticos.append(p)

    def generar_matriz(self, n):
        self.matrizPoliticos.clear()
        self.k = math.ceil(math.sqrt(n))
        self.m = math.ceil(n / self.k)
        total = self.k * self.m
        politicos = []

        for i in range(n):
            edad = random.randint(30, 79)
            p = Politico(i + 1, random.randint(100, 1000000), edad, str(2024 - edad))
            politicos.append(p)

        while len(politicos) < total:
            p = Politico(-1, -1, -1, "X")
            politicos.append(p)

        index = 0
        for i in range(self.k):
            fila = []
            for j in range(self.m):
                fila.append(politicos[index])
                index += 1
            self.matrizPoliticos.append(fila)

    def ordenar_array(self, algoritmo):
        if algoritmo == 1:
            self.bubble_sort(self.arrayPoliticos, False)  # Ordenar por dinero robado
        else:
            self.selection_sort(self.arrayPoliticos, False)  # Ordenar por dinero robado

    def ordenar_matriz(self, algoritmo):
        for i in range(len(self.matrizPoliticos)):
            if algoritmo == 1:
                self.bubble_sort(self.matrizPoliticos[i], False)
            else:
                self.selection_sort(self.matrizPoliticos[i], False)

        for j in range(self.m):
            for i in range(self.k - 1):
                min_index = i
                for l in range(i + 1, self.k):
                    if self.matrizPoliticos[l][j].edad < self.matrizPoliticos[min_index][j].edad:
                        min_index = l
                self.matrizPoliticos[i][j], self.matrizPoliticos[min_index][j] = self.matrizPoliticos[min_index][j], self.matrizPoliticos[i][j]

    def bubble_sort(self, vec, ordenar_por_edad):
        for i in range(len(vec) - 1):
            for j in range(len(vec) - i - 1):
                if (ordenar_por_edad and vec[j].edad > vec[j + 1].edad) or \
                   (not ordenar_por_edad and vec[j].dineroRobado > vec[j + 1].dineroRobado):
                    vec[j], vec[j + 1] = vec[j + 1], vec[j]

    def selection_sort(self, vec, ordenar_por_edad):
        for i in range(len(vec) - 1):
            min_index = i
            for j in range(i + 1, len(vec)):
                if (ordenar_por_edad and vec[j].edad < vec[min_index].edad) or \
                   (not ordenar_por_edad and vec[j].dineroRobado < vec[min_index].dineroRobado):
                    min_index = j
            vec[i], vec[min_index] = vec[min_index], vec[i]

File: Modelos Java/Politicos/src/main/java/com/mycompany/politicos/Modelo/Modelo.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.mycompany.politicos.Modelo;

/**
 *
 * @author 57320
 */
import java.util.*;

public class Modelo {
    public List<Politico> arrayPoliticos = new ArrayList<>();
    public List<List<Politico>> matrizPoliticos = new ArrayList<>();
    public int k, m;

    public void generarArray(int n) {
        arrayPoliticos.clear();
        Random rand = new Random();
        for (int i = 0; i < n; i++) {
            int edad = rand.nextInt(50) + 30;
            Politico p = new Politico(i + 1, rand.nextInt(1000000) + 100, edad, String.valueOf(2024 - edad));
            arrayPoliticos.add(p);
        }
    }

    public void generarMatriz(int n) {
        matrizPoliticos.clear();
        Random rand = new Random();
        k = (int) Math.ceil(Math.sqrt(n));
        m = (int) Math.ceil((double) n / k);
        int total = k * m;

        List<Politico> politicos = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            int edad = rand.nextInt(50) + 30;
            politicos.add(new Politico(i + 1, rand.nextInt(1000000) + 100, edad, String.valueOf(2024 - edad)));
        }

        while (politicos.size() < total) {
            politicos.add(new Politico());
        }

        int index = 0;
        for (int i = 0; i < k; i++) {
            List<Politico> fila = new ArrayList<>();
            for (int j = 0; j < m; j++) {
                fila.add(politicos.get(index++));
            }
            matrizPoliticos.add(fila);
        }
    }

    public void ordenarArray(int algoritmo) {
        if (algoritmo == 1)
            bubbleSort(arrayPoliticos, false); // Ordenar por dinero robado
        else
            selectionSort(arrayPoliticos, false);
    }

    public void ordenarMatriz(int algoritmo) {
        for (List<Politico> fila : matrizPoliticos) {
            if (algoritmo == 1)
                bubbleSort(fila, false);
            else
                selectionSort(fila, false);
        }

        for (int j = 0; j < m; j++) {
            for (int i = 0; i < k - 1; i++) {
                int minIndex = i;
                for (int l = i + 1; l < k; l++) {
                    if (matrizPoliticos.get(l).get(j).edad < matrizPoliticos.get(minIndex).get(j).edad) {
                        minIndex = l;
                    }
                }
                Politico temp = matrizPoliticos.get(i).get(j);
                matrizPoliticos.get(i).set(j, matrizPoliticos.get(minIndex).get(j));
                matrizPoliticos.get(minIndex).set(j, temp);
            }
        }
    }

    private void bubbleSort(List<Politico> lista, boolean ordenarPorEdad) {
        int n = lista.size();
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                Politico a = lista.get(j);
                Politico b = lista.get(j + 1);
                if ((ordenarPorEdad && a.edad > b.edad) || (!ordenarPorEdad && a.dineroRobado > b.dineroRobado)) {
                    lista.set(j, b);
                    lista.set(j + 1, a);
                }
            }
        }
    }

    private void selectionSort(List<Politico> lista, boolean ordenarPorEdad) {
        int n = lista.size();
        for (int i = 0; i < n - 1; i++) {
            int minIdx = i;
            for (int j = i + 1; j < n; j++) {
                Politico a = lista.get(j);
                Politico b = lista.get(minIdx);
                if ((ordenarPorEdad && a.edad < b.edad) || (!ordenarPorEdad && a.dineroRobado < b.dineroRobado)) {
                    minIdx = j;
                }
            }
            Collections.swap(lista, i, minIdx);
        }
    }
    
    private void insertionSort() {
        int n = politicos.size();
        for (int i = 1; i < n; i++) {
            Politico aux = politicos.get(i);
            int j = i - 1;

            while (j >= 0 && politicos.get(j).dineroRobado > aux.dineroRobado) {
                politicos.set(j + 1, politicos.get(j));
                j--;
            }
            politicos.set(j + 1, aux);
        }
    }
}
